### PR TITLE
Dontdrophandoffs

### PR DESCRIPTION
**Pull request by zzlk**

### CODE REVIEW

**Comment by zzlk** [comment]
```

fork_join/hydroflow_builder
                        time:   [8.5478 ms 8.6249 ms 8.7166 ms]
                        change: [-42.185% -41.607% -40.955%] (p = 0.00 < 0.05)
                        Performance has improved.

fork_join/hydroflow     time:   [8.1951 ms 8.2689 ms 8.3578 ms]
                        change: [-24.178% -23.398% -22.509%] (p = 0.00 < 0.05)
                        Performance has improved.

fan_out/hydroflow/scheduled
                        time:   [79.653 ms 80.178 ms 80.804 ms]
                        change: [-20.744% -20.210% -19.520%] (p = 0.00 < 0.05)
                        Performance has improved.

identity/hydroflow      time:   [19.152 ms 19.377 ms 19.645 ms]
                        change: [-19.510% -18.130% -16.635%] (p = 0.00 < 0.05)
                        Performance has improved.

micro/ops/join          time:   [107.63 µs 109.62 µs 112.42 µs]
                        change: [-21.186% -17.624% -14.069%] (p = 0.00 < 0.05)
                        Performance has improved.

identity/iter           time:   [6.5453 ms 6.5623 ms 6.5802 ms]
                        change: [-19.665% -12.086% -4.8412%] (p = 0.00 < 0.05)
                        Performance has improved.

micro/ops/next_tick/small
                        time:   [95.959 µs 96.715 µs 98.016 µs]
                        change: [-12.273% -10.793% -9.1170%] (p = 0.00 < 0.05)
                        Performance has improved.

fan_in/hydroflow        time:   [137.13 ms 137.89 ms 138.80 ms]
                        change: [-9.9853% -9.4744% -8.9448%] (p = 0.00 < 0.05)
                        Performance has improved.

micro/ops/merge         time:   [27.373 µs 28.860 µs 30.324 µs]
                        change: [-13.297% -7.5912% -1.4869%] (p = 0.02 < 0.05)
                        Performance has improved.

micro/ops/tee           time:   [6.4373 µs 6.5496 µs 6.7229 µs]
                        change: [-10.222% -7.9847% -5.7548%] (p = 0.00 < 0.05)
                        Performance has improved.

reachability/hydroflow/scheduled
                        time:   [2.4033 ms 2.4268 ms 2.4547 ms]
                        change: [-8.8046% -7.5119% -6.3377%] (p = 0.00 < 0.05)
                        Performance has improved.

fan_in/hydroflow/surface
                        time:   [109.94 ms 110.77 ms 111.73 ms]
                        change: [-7.9116% -6.9690% -6.0031%] (p = 0.00 < 0.05)
                        Performance has improved.

micro/ops/unique        time:   [265.02 µs 266.68 µs 268.78 µs]
                        change: [-8.5949% -6.2672% -4.0311%] (p = 0.00 < 0.05)
                        Performance has improved.

micro/ops/map           time:   [9.1443 µs 9.1883 µs 9.2487 µs]
                        change: [-6.1354% -4.3649% -2.6150%] (p = 0.00 < 0.05)
                        Performance has improved.

reachability/hydroflow/surface_cheating
                        time:   [1.8849 ms 1.8980 ms 1.9141 ms]
                        change: [-5.0144% -4.3348% -3.4911%] (p = 0.00 < 0.05)
                        Performance has improved.

micro/ops/sort          time:   [137.75 µs 139.35 µs 141.35 µs]
                        change: [-5.3845% -3.2987% -1.1691%] (p = 0.00 < 0.05)
                        Performance has improved.

micro/ops/crossjoin     time:   [225.62 µs 225.84 µs 226.10 µs]
                        change: [-3.5578% -3.0947% -2.7019%] (p = 0.00 < 0.05)
                        Performance has improved.

upcase_in_place/iter    time:   [4.4883 ms 4.4949 ms 4.5020 ms]
                        change: [-4.5762% -2.9951% -1.7406%] (p = 0.00 < 0.05)
                        Performance has improved.

fan_out/sol             time:   [5.1356 ms 5.1895 ms 5.2448 ms]
                        change: [-4.0104% -2.7645% -1.5142%] (p = 0.00 < 0.05)
                        Performance has improved.















micro/ops/flat_map      time:   [18.419 µs 18.904 µs 19.691 µs]
                        change: [+10.611% +13.721% +17.817%] (p = 0.00 < 0.05)
                        Performance has regressed.

reachability/hydroflow  time:   [2.8089 ms 2.9047 ms 3.0077 ms]
                        change: [+7.0714% +11.083% +15.689%] (p = 0.00 < 0.05)
                        Performance has regressed.

fork_join/raw           time:   [6.2749 ms 6.3341 ms 6.3983 ms]
                        change: [+4.4761% +5.4348% +6.5777%] (p = 0.00 < 0.05)
                        Performance has regressed.

concatting/raw          time:   [51.164 ms 52.239 ms 53.544 ms]
                        change: [+3.1501% +5.3298% +8.0809%] (p = 0.00 < 0.05)
                        Performance has regressed.

arithmetic/iter-collect time:   [6.8125 ms 6.9027 ms 7.0223 ms]
                        change: [+1.8627% +4.6854% +7.3189%] (p = 0.00 < 0.05)
                        Performance has regressed.

fan_in/timely           time:   [70.492 ms 71.540 ms 72.656 ms]
                        change: [+2.7099% +4.6212% +6.5469%] (p = 0.00 < 0.05)
                        Performance has regressed.

join/String/String/timely
                        time:   [81.578 ms 82.573 ms 83.710 ms]
                        change: [+2.9498% +4.1747% +5.5800%] (p = 0.00 < 0.05)
                        Performance has regressed.

symmetric_hash_join/no_match
                        time:   [111.24 µs 111.95 µs 112.84 µs]
                        change: [+1.9744% +4.0784% +6.4715%] (p = 0.00 < 0.05)
                        Performance has regressed.

micro/ops/next_tick/big time:   [85.477 µs 86.422 µs 87.748 µs]
                        change: [+1.7521% +3.5310% +5.5341%] (p = 0.00 < 0.05)
                        Performance has regressed.

arithmetic/raw          time:   [19.203 ms 19.373 ms 19.554 ms]
                        change: [+2.2060% +3.2287% +4.1659%] (p = 0.00 < 0.05)
                        Performance has regressed.

arithmetic/hydroflow/compiled_no_cheating
                        time:   [6.0051 ms 6.1014 ms 6.2304 ms]
                        change: [+1.3430% +2.9366% +5.2247%] (p = 0.00 < 0.05)
                        Performance has regressed.

join/String/String/sol  time:   [76.052 ms 77.001 ms 78.023 ms]
                        change: [+1.0034% +2.3739% +3.9680%] (p = 0.00 < 0.05)
                        Performance has regressed.
```

File: hydroflow_lang/src/graph/hydroflow_graph.rs [hunk 1]
```diff
@@ -561,7 +561,13 @@ impl HydroflowGraph {
 
                 let recv_port_code = recv_ports
                     .iter()
-                    .map(|ident| quote! { let #ident = #ident.take_inner().into_iter(); });
+                    .map(|ident| {
+                        let borrow_ident = Ident::new(&*format!("{}_borrow", ident), ident.span());
+                        quote! {
+                            let mut #borrow_ident = #ident.borrow_mut_swap();
+                            let #ident = #borrow_ident.drain(..);
+                        }
```
Comment (MingweiSamuel) [inline hydroflow_lang/src/graph/hydroflow_graph.rs:569]: Would be fine to just reuse `#ident`, but doesn't really matter

**Review by MingweiSamuel — APPROVED** [review]
Nice very juicy PR

**Comment by MingweiSamuel** [comment]
Need to update the comment `A [VecDeque]-based FIFO handoff.`

**Comment by MingweiSamuel** [comment]
#480

File: hydroflow_lang/src/graph/hydroflow_graph.rs (hunk 1 above)
Comment (zzlk) [reply hydroflow_lang/src/graph/hydroflow_graph.rs:569]: like
let #ident = <>;
let #ident = #ident.drain(..);
?

**Pull request merged**